### PR TITLE
PE-1885: Implements a function to derive the IPFS hash

### PR DESCRIPTION
**Pull request by matibat**
Changes:
- Implements the `deriveIpfsCid` function that is a wrapper for the `only-ipfs-hash` package
   - Adds unit tests
- Declares the `--add-ipfs-tag` flag parameter
- Adds an entry to the README

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -764,6 +764,16 @@ NOTE: To upload to the root of a drive, specify its root folder ID as the parent
...
 
+### The IPFS+Arweave integration
+
+If you'd like to share the file you are attempting to upload to the IPFS protocol as well, you can set the `--ipfs` flag to the `upload-file` command.
+
+```shell
+ardrive upload-file --ipfs --local-path /path/to/file.txt  --parent-folder-id "9af694f6-4cfc-4eee-88a8-1b02704760c0" -w /path/to/wallet.json
+```
+
+By using the flag you are making use of the `IPFS+Arweave` ecosystem. You can find more information about the integration [here](https://arweave.medium.com/arweave-ipfs-persistence-for-the-interplanetary-file-system-9f12981c36c3);
```
Comment (matibat) [inline README.md:775]: Is it OK to have this link here?

File: src/utils/derivate_ipfs_cid.test.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { expect } from 'chai';
+import { derivateIpfsCid } from './derivate_ipfs_cid';
+
+const dataToHash = [
+	{ data: 'hello', hash: 'QmWfVY9y3xjsixTgbd9AorQxH7VtMpzfx2HaWtsoUYecaX' },
+	{ data: 'my file data', hash: 'Qmd8s28wFUPP2rRbQEvvoHh84G22ShP5YzNgvT2mdhyrmq' },
+	{ data: Buffer.from('my file data'), hash: 'Qmd8s28wFUPP2rRbQEvvoHh84G22ShP5YzNgvT2mdhyrmq' }
+];
+
+describe('derivateIpfsCid function', () => {
```
Comment (matibat) [inline src/utils/derivate_ipfs_cid.test.ts:10]: Q: Is there anything else to unit test here?

File: src/utils/derivate_ipfs_cid.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const CID = require('ipfs-only-hash');
```
Comment (matibat) [inline src/utils/derivate_ipfs_cid.ts:2]: It didn't let me use `import` :b

**Review by matibat — COMMENTED** [review]
\* self-review checkpoint *

File: src/parameter_declarations.ts [hunk 4]
```diff
@@ -466,3 +468,10 @@ Parameter.declare({
 	description:
 		'(OPTIONAL) Provide a custom content type to all files within the upload to be used by the gateway to display the content'
 });
+
+Parameter.declare({
```
Comment (matibat) [inline src/parameter_declarations.ts:472]:
```suggestion
// TODO: double check with the team the flag name and description
Parameter.declare({
```

File: src/utils/derive_ipfs_cid.test.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { expect } from 'chai';
+import { deriveIpfsCid } from './derive_ipfs_cid';
+
+const dataToHash = [
+	{ data: 'hello', hash: 'QmWfVY9y3xjsixTgbd9AorQxH7VtMpzfx2HaWtsoUYecaX' },
+	{ data: 'my file data', hash: 'Qmd8s28wFUPP2rRbQEvvoHh84G22ShP5YzNgvT2mdhyrmq' },
+	{ data: Buffer.from('my file data'), hash: 'Qmd8s28wFUPP2rRbQEvvoHh84G22ShP5YzNgvT2mdhyrmq' }
+];
+
+describe('derivateIpfsCid function', () => {
```
Comment (matibat) [inline src/utils/derive_ipfs_cid.test.ts:10]:
```suggestion
describe('deriveIpfsCid function', () => {
```

File: src/utils/derive_ipfs_cid.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const CID = require('ipfs-only-hash');
```
Comment (arielmelendez) [inline src/utils/derive_ipfs_cid.ts:2]: Consider a more generic name for this file like `ipfs_utils.ts`. Naming it for a single function is very limiting.

File: src/utils/derive_ipfs_cid.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+
+type DataInput = string | Uint8Array | AsyncIterable<Uint8Array>;
```
Comment (arielmelendez) [inline src/utils/derive_ipfs_cid.ts:4]: Consider type name `CIDHashable` - `DataInput` is wildly generic.

File: src/utils/derive_ipfs_cid.test.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+const dataToHash = [
+	{ data: 'hello', hash: 'QmWfVY9y3xjsixTgbd9AorQxH7VtMpzfx2HaWtsoUYecaX' },
+	{ data: 'my file data', hash: 'Qmd8s28wFUPP2rRbQEvvoHh84G22ShP5YzNgvT2mdhyrmq' },
+	{ data: Buffer.from('my file data'), hash: 'Qmd8s28wFUPP2rRbQEvvoHh84G22ShP5YzNgvT2mdhyrmq' }
+];
+
+describe('deriveIpfsCid function', () => {
+	it('returns the expeced hash', async () => {
+		for (const expected of dataToHash) {
```
Comment (arielmelendez) [inline src/utils/derive_ipfs_cid.test.ts:12]: Consider the variable name `testCaseData` over `expected`.

File: src/utils/derive_ipfs_cid.test.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import { expect } from 'chai';
+import { deriveIpfsCid } from './derive_ipfs_cid';
+
+const dataToHash = [
+	{ data: 'hello', hash: 'QmWfVY9y3xjsixTgbd9AorQxH7VtMpzfx2HaWtsoUYecaX' },
```
[the remark this answers is not in the document]
Comment (arielmelendez) [reply src/utils/derive_ipfs_cid.test.ts:5]: Also consider `expectedHash` over `hash` for enhanced test readability.

**Review by arielmelendez — CHANGES_REQUESTED** [review]
Made some input flag and file and variable naming requests and suggestions.

File: README.md [hunk 10]
```diff
@@ -764,6 +764,16 @@ NOTE: To upload to the root of a drive, specify its root folder ID as the parent
 ardrive drive-info -d "c7f87712-b54e-4491-bc96-1c5fa7b1da50" | jq -r '.rootFolderId'
 ```
 
+### The IPFS+Arweave integration
+
+If you'd like to share the file you are attempting to upload to the IPFS protocol as well, you can set the `--ipfs` flag to the `upload-file` command.
```
Comment (karlprieb) [inline README.md:769]: That's not true as we're simply adding tags that can be used when querying gql

File: src/parameter_declarations.ts [hunk 11]
```diff
@@ -41,6 +41,7 @@ export const LocalCSVParameter = 'localCsv';
 export const WithKeysParameter = 'withKeys';
 export const GatewayParameter = 'gateway';
 export const CustomContentTypeParameter = 'contentType';
+export const IPFSParameter = 'ipfs';
```
[the remark this answers is not in the document]
Comment (karlprieb) [reply src/parameter_declarations.ts:44]: `--add-ipfs-tag` is my favorite 😄

File: README.md (hunk 10 above)
Comment (matibat) [reply README.md:769]: > simply adding tags that can be used when querying gql

And for the IPFS+Arweave nodes to work as a pinning service. So, do we want to explain the part of the GQL tag to the user or by having the link there is just enough?

File: src/parameter_declarations.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (matibat) [reply src/parameter_declarations.ts]: There: 035d59c47d16e79b7de9b3b6f59a6afd870e3670

File: src/utils/derive_ipfs_cid.ts (hunk 7 above)
Comment (matibat) [reply src/utils/derive_ipfs_cid.ts]: f5f33ce4fdf7052d19cdf66f14fa55e45a7ed29b

File: src/utils/derive_ipfs_cid.test.ts (hunk 8 above)
Comment (matibat) [reply src/utils/derive_ipfs_cid.test.ts]: 8d348aa8aeb59d755388ae72a0415036253cfcce

File: src/utils/derive_ipfs_cid.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (matibat) [reply src/utils/derive_ipfs_cid.test.ts]: 8d348aa8aeb59d755388ae72a0415036253cfcce

**Comment by matibat** [comment]
TODO:
- Update the README.md
- Update the flag description

**Comment by matibat** [comment]
The previous commit makes the flag allowed solely for public entities

File: src/utils/derive_ipfs_cid.ts (hunk 6 above)
Comment (arielmelendez) [reply src/utils/derive_ipfs_cid.ts]: pretty good but the test file name should probably match

File: src/parameter_declarations.ts [hunk 12]
```diff
@@ -466,3 +468,13 @@ Parameter.declare({
 	description:
 		'(OPTIONAL) Provide a custom content type to all files within the upload to be used by the gateway to display the content'
 });
+
+// TODO: double check with the team the flag name and description
+Parameter.declare({
+	name: IPFSParameter,
+	aliases: ['--add-ipfs-tag'],
+	description:
+		'(OPTIONAL) Computes the v1 IPFS content identifier (CID) for each file and sets it as the "IPFS-Add" tag value of each\'s respective ArFS metadata transaction',
```
Comment (arielmelendez) [inline src/parameter_declarations.ts:477]:
```suggestion
		'(OPTIONAL) Computes the v1 IPFS content identifier (CID) for each file and sets it as the "IPFS-Add" tag value of each\'s respective file data transaction',
```

**Review by arielmelendez — CHANGES_REQUESTED** [review]
Left a suggestion for the README. Also consider renaming the derive_ipfs_cid.test.ts file to ipfs_utils.ts.

**Review by arielmelendez — APPROVED** [review]

**Pull request merged**